File: src/main/java/net/database/WishlistAPI.java
package net.database;

import client.Character;
import org.jooq.Record;
import org.jooq.Result;

import java.util.*;

import static database.jooq.Tables.WISHLIST;

public class WishlistAPI {

    public static void load(Character chr) {
        List<Integer> list = getWishList(chr.getId());
        for (int i = 0; i < 10; i++) {
            if (list.size() <= i) {
                chr.getWishlist()[i] = 0;
            } else {
                chr.getWishlist()[i] = list.get(i);
            }
        }
    }

    public static void save(Character chr) {
        List<Integer> old = getWishList(chr.getId());

        Arrays.stream(chr.getWishlist()).forEach(wish -> {
            if (wish != 0) {
                if (!old.contains(wish)){ // new wish
                    add(chr.getId(), wish);
                } else {
                    old.remove(Integer.valueOf(wish));
                }
            }
        });

        // old wishes that are not in current wishlist
        old.forEach(wish -> remove(chr.getId(), wish));
    }

    private static List<Integer> getWishList(int cid) {
        Result<Record> res = DatabaseCore.getConnection()
                .select().from(WISHLIST)
                .where(WISHLIST.CID.eq(cid))
                .fetch();
        List<Integer> wishlist = new ArrayList<>();
        res.forEach(rec -> wishlist.add(rec.getValue(WISHLIST.SN)));

        return wishlist;
    }

    private static void add(int cid, int sn) {
        DatabaseCore.getConnection()
                .insertInto(WISHLIST, WISHLIST.CID, WISHLIST.SN)
                .values(cid, sn)
                .execute();
    }

    private static void remove(int cid, int sn) {
        DatabaseCore.getConnection()
                .deleteFrom(WISHLIST)
                .where(WISHLIST.CID.eq(cid))
                .and(WISHLIST.SN.eq(sn))
                .execute();
    }
}

File: src/main/java/net/database/WishlistAPI.kt
package net.database

import client.Character
import database.jooq.Tables.WISHLIST
import net.database.DatabaseCore.connection
import java.util.*

object WishlistAPI {

    fun load(chr: Character) {
        val list: List<Int> = getWishList(chr.id)
        for (i in 0..9) {
            if (list.size <= i) {
                chr.wishlist[i] = 0
            } else {
                chr.wishlist[i] = list[i]
            }
        }
    }

    fun save(chr: Character) {
        val old = getWishList(chr.id)

        Arrays.stream(chr.wishlist).forEach {
            if (it != 0) {
                if (!old.contains(it)) { // new wish
                    add(chr.id, it)
                } else {
                    old.remove(Integer.valueOf(it))
                }
            }
        }

        // old wishes that are not in current wishlist
        old.forEach { remove(chr.id, it) }
    }

    private fun getWishList(cid: Int): MutableList<Int> {
        with(WISHLIST) {
            val res = connection.select().from(this)
                .where(CID.eq(cid))
                .fetch()

            val wishlist: MutableList<Int> = ArrayList()
            res.forEach { wishlist.add(it.getValue(SN)) }
            return wishlist
        }
    }

    private fun add(cid: Int, sn: Int) {
        with(WISHLIST) {
            connection.insertInto(this, CID, SN)
                .values(cid, sn)
                .execute()
        }
    }

    private fun remove(cid: Int, sn: Int) {
        with(WISHLIST) {
            connection.deleteFrom(this)
                .where(CID.eq(cid))
                .and(SN.eq(sn))
                .execute()
        }
    }
}